File: 1707-Maximize-Xor/1707.py
class Trie:
  L = 30

  def __init__(self):
    self.left = None
    self.right = None
    self.min_value = float("inf")

  def insert(self, val: int):
    node = self
    node.min_value = min(node.min_value, val)
    for i in range(Trie.L, -1, -1):
      bit = (val >> i) & 1
      if bit == 0:
        if not node.left:
          node.left = Trie()
        node = node.left
      else:
        if not node.right:
          node.right = Trie()
        node = node.right
      node.min_value = min(node.min_value, val)

  def getMaxXorWithLimit(self, val: int, limit: int) -> int:
    node = self
    if node.min_value > limit:
      return - 1
      
    ans = 0
    for i in range(Trie.L, -1, -1):
      bit = (val >> i) & 1
      check = False
      if bit == 0:
        if node.right and node.right.min_value <= limit:
          node = node.right
          check = True
        else:
          node = node.left
      else:
        if node.left and node.left.min_value <= limit:
          node = node.left
          check = True
        else:
          node = node.right
      if check:
        ans |= 1 << i
    return ans

class Solution:
  def maximizeXor(self, nums: List[int], queries: List[List[int]]) -> List[int]:
    t = Trie()
    for val in nums:
      t.insert(val)

    q = len(queries)
    ans = [0] * q
    for i, (x, m) in enumerate(queries):
      ans[i] = t.getMaxXorWithLimit(x, m)

    return ans

File: 1707-Maximize-Xor/1707.java
class Solution {
  public int[] maximizeXor(int[] nums, int[][] queries) {
    Trie trie = new Trie();
    for (int val : nums) {
      trie.insert(val);
    }
    int numQ = queries.length;
    int[] ans = new int[numQ];
    for (int i = 0; i < numQ; ++i) {
      ans[i] = trie.getMaxXorWithLimit(queries[i][0], queries[i][1]);
    }
    return ans;
  }
}

class Trie {
  static final int L = 30;
  Trie[] children = new Trie[2];
  int min = Integer.MAX_VALUE;

  public void insert(int val) {
    Trie node = this;
    node.min = Math.min(node.min, val);
    for (int i = L - 1; i >= 0; --i) {
      int bit = (val >> i) & 1;
      if (node.children[bit] == null) {
        node.children[bit] = new Trie();
      }
      node = node.children[bit];
      node.min = Math.min(node.min, val);
    }
  }

  public int getMaxXorWithLimit(int val, int limit) {
    Trie node = this;
    if (node.min > limit) {
      return -1;
    }
    int ans = 0;
    for (int i = L - 1; i >= 0; --i) {
      int bit = (val >> i) & 1;
      if (node.children[bit ^ 1] != null && node.children[bit ^ 1].min <= limit) {
        ans |= 1 << i;
        bit ^= 1;
      }
      node = node.children[bit];
    }
    return ans;
  }
}